### PR TITLE
BL-2911 Undo for Origami

### PR DESCRIPTION
**Pull request by JohnThomson**
[Review on Reviewable](https://reviewable.io/reviews/bloombooks/bloomdesktop/902)

### CODE REVIEW

File: src/BloomBrowserUI/bookEdit/js/calledByCSharp.js [hunk 1]
```diff
@@ -2,10 +2,14 @@ var CalledByCSharp = (function () {
     function CalledByCSharp() {
     }
     CalledByCSharp.prototype.handleUndo = function () {
+        var contentWindow = this.getPageContent()
+        if (contentWindow && contentWindow.origamiCanUndo()) {
+            contentWindow.origamiUndo();
+        }
         // Stuff "in the toolbox" (not clear what that means) gets its own undo handling
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/js/calledByCSharp.js]: Are we clear what "in the toolbox" means, now?

File: src/BloomBrowserUI/bookEdit/js/origami.js [hunk 2]
```diff
@@ -65,31 +76,79 @@ function layoutToggleClickHandler() {
...
-    newSplitPane.prepend(getSplitPaneDividerHtml(verticalOrHorizontal));
-    newSplitPane.prepend(getSplitPaneComponentWithNewContent(newContentPosition));
-    newSplitPane.splitPane();
+var origamiUndoStack = [];
+var origamiUndoIndex = 0; // of item that should be redone next, if any
+
+// I made this take an argument of the closest parent that survies the change with only
+// its children modified. I could not make it work for multiple levels of Undo and Redo
+// (including deletions) by only cloning that element, as I first attempted; but I
+// decided to keep the argument in case something requires us to attempt that again one day.
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/js/origami.js]: But in the meantime it made me misunderstand all the other code that was calling this :-)

File: src/BloomBrowserUI/bookEdit/js/origami.js [hunk 3]
```diff
@@ -65,31 +76,79 @@ function layoutToggleClickHandler() {
...
+}
+
+function origamiRedo() {
+    if (origamiCanRedo()) {
+        origamiUndoIndex++;
+        $('.marginBox').replaceWith(origamiUndoStack[origamiUndoIndex].original);
+    }
+}
+
+function undoSplit(innerElement) {
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/js/origami.js]: I can't see from where this is called.

File: src/BloomBrowserUI/bookEdit/js/origami.js [hunk 4]
```diff
@@ -65,31 +76,79 @@ function layoutToggleClickHandler() {
...
-    newSplitPane.splitPane();
+var origamiUndoStack = [];
+var origamiUndoIndex = 0; // of item that should be redone next, if any
+
+// I made this take an argument of the closest parent that survies the change with only
+// its children modified. I could not make it work for multiple levels of Undo and Redo
+// (including deletions) by only cloning that element, as I first attempted; but I
+// decided to keep the argument in case something requires us to attempt that again one day.
+function addUndoItem(parentElement) {
+    origamiUndoStack.length = origamiUndoIndex; // truncate any redo items
```
Comment (hatton) [inline src/BloomBrowserUI/bookEdit/js/origami.js]: So in light of what's it's doing... is a better name "addUndoPoint"  (as in "point in time")?

**Comment by JohnThomson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/js/calledByCSharp.js, line 9 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902#-K5lwvbLvDSnXaYrothM-r1-9:-K5mA0CSdMPWtt4WTnAa:1474230589) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/52bd5ba78a8fc208e728d29613cae02cbb3f0a49/src/BloomBrowserUI/bookEdit/js/calledByCSharp.js#L9)):*
I think so. Updated comments.

---

*[src/BloomBrowserUI/bookEdit/js/origami.js, line 109 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902#-K5lwvbLvDSnXaYrothN-r1-109:-K5m6_4pjTztT7QF79R2:-1863343418) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/52bd5ba78a8fc208e728d29613cae02cbb3f0a49/src/BloomBrowserUI/bookEdit/js/origami.js#L109)):*
OK, I removed them.

---

*[src/BloomBrowserUI/bookEdit/js/origami.js, line 111 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902#-K5lwvbLvDSnXaYrothN-r1-111:-K5m6ZaI6ySElrydYX6o:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/52bd5ba78a8fc208e728d29613cae02cbb3f0a49/src/BloomBrowserUI/bookEdit/js/origami.js#L111)):*
Done.

---

*[src/BloomBrowserUI/bookEdit/js/origami.js, line 148 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902#-K5lwvbLvDSnXaYrothN-r1-148:-K5m5PlORl1tq7SsqAEE:207042611) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/52bd5ba78a8fc208e728d29613cae02cbb3f0a49/src/BloomBrowserUI/bookEdit/js/origami.js#L148)):*
It isn't. Part of an approach that didn't work out. Removed it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902)*

**Comment by hatton** [comment]
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/902)*

**Pull request merged**